main/contracts/index.ts

```diff
@@ -44,45 +29,16 @@ function parseAbi (abiData: string): Interface | undefined {
   }
 }
 
-// function scanEndpoint (contractAddress: Address, chainId: string) {
-//   if (chainId === '0x1') {
-//     return `https://api.etherscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=3SYU5MW5QK8RPCJV1XVICHWKT774993S24`
-//   } else if (chainId === '0x89') {
-//     return `https://api.polygonscan.com/api?module=contract&action=getsourcecode&address=0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270&apikey=YourApiKeyToken`
-//   } else if ('optimism') {
-//     return `https://api-optimistic.etherscan.io/api?module=contract&action=getsourcecode&address=0x80AA7cb0006d5DDD91cce684229Ac6e398864606&apikey=YourApiKeyToken`
-//   } else if ('arbitrum') {
-//     return `https://api.arbiscan.io/api?module=contract&action=getsourcecode&address=0x0000000000000000000000000000000000001004&apikey=YourApiKeyToken`
-//   }
-// }
-
-async function fetchSourceCode (contractAddress: Address, chainId: string) {
-  const res = await fetch(`https://api.etherscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=3SYU5MW5QK8RPCJV1XVICHWKT774993S24`)
-
-  if (res.status === 200 && (res.headers.get('content-type') || '').toLowerCase().includes('json')) {
-    const parsedResponse = await (res.json() as Promise<EtherscanSourceCodeResponse>)
-
-    if (parsedResponse.message === 'OK') return parsedResponse.result
-  }
-
-  return []
-}
-
-async function fetchAbi (contractAddress: Address, chainId: string): Promise<ContractSourceCodeResult | undefined> {
+export async function fetchContract (contractAddress: Address, chainId: string): Promise<ContractSource | undefined> {
   try {
-    const sources = await fetchSourceCode(contractAddress, chainId)
-
-    if (sources.length > 0) {
-      const source = sources[0]
-      const implementation = source.Implementation
-
-      if (implementation) {
-        // this is a proxy contract, return the ABI for the source
-        return fetchAbi(implementation, chainId)
-      }
-
-      return source
-    }
+    const [sourcifyContract, scanContract] = await Promise.all([fetchSourcifyContract(contractAddress, chainId), fetchScanContract(contractAddress, chainId)])
+    
+    // if no scan result and scan supports the chain, return undefined
+    if (scanContract && scanContract.abi === 'Contract source code not verified' && chainSupported(chainId)) {
```

Comment:
any Etherscan-specific code should go in the Etherscan module and we should reflect these types of possible responses in the returned interface

Reply:
also what's the logic in checking this value first but then returning the sourcify contract first by default?

Reply:
Quoting Jordan: 
"First, I think think our preference is to prioritize sourcify as a higher quality abi source due to it being an EF project that seems to care about decentralized access, hosting everything on IPFS -- in the future we could do this lookup via IPFS instead

Second, in a case where the abi is on sourcify but not on etherscan, but etherscan supports that chain, our result would be not finding the abi

To address this we could try to get the abi from both sources and then make some decisions about which to return based on priority logic"

Reply:
Not sure how you want this particular logic to be architected, personally I think it's ok like this, feel free to suggest otherwise or put in a commit.

Reply:
I think Jordan was just enumerating the reasons why the first implementation wasn't quite right. I don't understand the logic here, why are we returning undefined if Etherscan doesn't have the abi but Sourcify does?

Reply:
What I was describing here was mainly problems with the initial implementation. The way I think the final implementation should work is: Look for ABI from both sources, if we get a result from Sourcify return it immediately, if we don't get a result from Sourcify or that request times out but we do find it on Etherscan then return the Etherscan version. Speed is important here so we want to return the best option as soon as possible -- and when don't find anything we should probably record that too so the UI can know we're done looking.

Reply:
Ok, that changes things somewhat.  Still got the index tests to rework in any case, so will take about the same time.